File: src/org/genyris/math/PlusFunction.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.math;

import java.math.BigDecimal;

import org.genyris.core.Bignum;
import org.genyris.core.Exp;
import org.genyris.core.Lsymbol;
import org.genyris.interp.Interpreter;

public class PlusFunction extends AbstractMathFunction {

    public PlusFunction(Interpreter interp, Lsymbol name) {
        super(interp, name, 1);
    }

    protected Exp mathOperation(Exp a) {
        return a;
    }
    protected Exp mathOperation(Exp a, Exp b) {
        return new Bignum(((BigDecimal) a.getJavaValue()).add((BigDecimal) b.getJavaValue()));
    }
}

File: src/org/genyris/math/PlusFunction.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.math

import org.genyris.core.Bignum
import org.genyris.core.Exp
import org.genyris.interp.Interpreter

class PlusFunction(interp: Interpreter?) : AbstractMathFunction(interp, "+", 1) {
    override fun mathOperation(a: Exp?): Exp? {
        return a
    }

    override fun mathOperation(a: Exp, b: Exp?): Exp? {
        return (a as Bignum).add(b as Bignum?)
    }
}
